### PR TITLE
[SPARK-8060] Improve DataFrame Python test coverage and documentation.

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

**Comment by rxin** [comment]
cc @davies

File: python/pyspark/sql/context.py [hunk 1]
```diff
@@ -196,7 +199,7 @@ def _inferSchema(self, rdd, samplingRatio=None):
                              "can not infer schema")
         if type(first) is dict:
             warnings.warn("Using RDD of dict to inferSchema is deprecated,"
-                          "please use pyspark.sql.Row instead")
+                          "please use pyspark.sql.Row instead", DeprecationWarning)
```
Comment (davies) [inline python/pyspark/sql/context.py]: DeprecationWarning is ignored by default.

File: python/pyspark/sql/context.py [hunk 2]
```diff
@@ -359,18 +365,12 @@ def registerDataFrameAsTable(self, df, tableName):
...
         """Loads a Parquet file, returning the result as a :class:`DataFrame`.
 
-        >>> import tempfile, shutil
-        >>> parquetFile = tempfile.mkdtemp()
-        >>> shutil.rmtree(parquetFile)
-        >>> df.saveAsParquetFile(parquetFile)
-        >>> df2 = sqlContext.parquetFile(parquetFile)
-        >>> sorted(df.collect()) == sorted(df2.collect())
-        True
+        .. note:: Deprecated in 1.4, use :func:`DataFrameReader.parquet` instead.
```
Comment (davies) [inline python/pyspark/sql/context.py]: This is a different implementation than data source api, may have different behavior. Even we deprecate it, should still have tests for it (make sure it's not broken).

File: python/pyspark/sql/context.py [hunk 3]
```diff
@@ -378,39 +378,12 @@ def parquetFile(self, *paths):
...
-        ...     StructField("field3",
-        ...         StructType([StructField("field5", ArrayType(IntegerType()))]))])
-        >>> df2 = sqlContext.jsonFile(jsonFile, schema)
-        >>> df2.printSchema()
-        root
-         |-- field2: string (nullable = true)
-         |-- field3: struct (nullable = true)
-         |    |-- field5: array (nullable = true)
-         |    |    |-- element: integer (containsNull = true)
+        .. note:: Deprecated in 1.4, use :func:`DataFrameReader.json` instead.
```
Comment (davies) [inline python/pyspark/sql/context.py]: Same as parquetFile().

File: python/pyspark/sql/dataframe.py [hunk 4]
```diff
@@ -653,9 +627,11 @@ def describe(self, *cols):
...
     @since(1.3)
     def head(self, n=None):
-        """
-        Returns the first ``n`` rows as a list of :class:`Row`,
-        or the first :class:`Row` if ``n`` is ``None.``
+        """Returns the first ``n`` rows.
+
+        If n is greater than 1, return a list of :class:`Row`. If n is 1, return a single Row.
+
+        :param n: int, default 1.
```
Comment (davies) [inline python/pyspark/sql/dataframe.py]: head() is different head(1), so we can not say `n` has `1` as default value.

File: python/run-tests [hunk 5]
```diff
@@ -70,13 +70,13 @@ function run_core_tests() {
 
 function run_sql_tests() {
     echo "Run sql tests ..."
+    run_test "pyspark.sql.readwriter"
```
Comment (davies) [inline python/run-tests]: Should this change be reverted after debugging?

File: python/run-tests [hunk 6]
```diff
@@ -142,11 +142,11 @@ fi
 echo "Testing with Python version:"
 $PYSPARK_PYTHON --version
 
-run_core_tests
+#run_core_tests
```
Comment (davies) [inline python/run-tests]: Also revert these.

File: python/pyspark/sql/context.py (hunk 1 above)
Comment (rxin) [inline python/pyspark/sql/context.py]: do you mean we should remove all DeprecationWarning?

**Comment by davies** [comment]
LGTM

**Pull request closed without merging**